library/oidc.c

```diff
@@ -0,0 +1,389 @@
+        http_resp->body_cb = parse_token_cb;
+    } else {
+        failed_auth_req(req, http_resp->status);
+    }
+}
+
+static void code_cb(tlsuv_http_resp_t *http_resp, void *ctx) {
+    auth_req *req = ctx;
+    if (http_resp->code / 100 == 3) {
+        const char *redirect = tlsuv_http_resp_header(http_resp, "Location");
+        struct tlsuv_url_s uri;
+        tlsuv_parse_url(&uri, redirect);
+        char *code = strstr(uri.query, "code=");
+        code += strlen("code=");
+
+        ZITI_LOG(DEBUG, "requesting token");
+        const char *path = get_endpoint_path(req->clt, "token_endpoint");
+        tlsuv_http_req_t *token_req = tlsuv_http_req(&req->clt->http, "POST", path, token_cb, req);
+        token_req->data = req;
+        tlsuv_http_req_form(token_req, 8, (tlsuv_http_pair[]) {
+                {"code",                  code},
+                {"grant_type",            "authorization_code"},
+                {"code_verifier",         req->code_verifier},
+                {"code_challenge",        req->code_challenge},
+                {"code_challenge_method", "S256"},
+                {"client_id",             req->clt->client_id},
+                {"scopes",                "openid offline_access"},
+                {"redirect_uri", default_cb_url}
+        });
+    } else {
+        failed_auth_req(req, http_resp->status);
+    }
+}
+
+static void login_cb(tlsuv_http_resp_t *http_resp, void *ctx) {
+    auth_req *req = ctx;
+    if (http_resp->code / 100 == 3) {
+        const char *redirect = tlsuv_http_resp_header(http_resp, "Location");
+        struct tlsuv_url_s uri;
+        tlsuv_parse_url(&uri, redirect);
+
+        tlsuv_http_req(&req->clt->http, "GET", uri.path, code_cb, req);
+    } else {
+        failed_auth_req(req, http_resp->status);
+    }
+}
+
+static void auth_cb(tlsuv_http_resp_t *http_resp, void *ctx) {
+    auth_req *req = ctx;
+    if (http_resp->code / 100 == 3) {
+        const char *redirect = tlsuv_http_resp_header(http_resp, "Location");
+        struct tlsuv_url_s uri;
+        tlsuv_parse_url(&uri, redirect);
+        char *p = strstr(uri.query, "authRequestID=");
+        p += strlen("authRequestID=");
+
+        ZITI_LOG(DEBUG, "logging in with cert auth");
+        tlsuv_http_req_t *login_req = tlsuv_http_req(&req->clt->http, "POST", "/oidc/login/cert", login_cb, req);
+        tlsuv_http_req_form(login_req, 1, &(tlsuv_http_pair) {"id", p});
+    } else {
+        failed_auth_req(req, http_resp->status);
+    }
+}
+
+int oidc_client_start(oidc_client_t *clt, oidc_token_cb cb) {
+    clt->token_cb = cb;
+    ZITI_LOG(DEBUG, "requesting authentication code");
+    auth_req *req = new_auth_req(clt);
+
+    const char *path = get_endpoint_path(clt, "authorization_endpoint");
+    tlsuv_http_req_t *http_req = tlsuv_http_req(&clt->http, "POST", path, auth_cb, req);
+    tlsuv_http_req_header(http_req, "Accept", "*/*");
+    int rc = tlsuv_http_req_form(http_req, 6, (tlsuv_http_pair[]) {
```

Comment:
The default OIDC implementation uses HTTP form submission but also supports JSON for convenience. Doesn't need to be changed here as form submissions must be supported.